server/src/test/java/org/elasticsearch/cluster/health/NoDataNodesHealthTests.java

```diff
@@ -0,0 +1,153 @@
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.RoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+import org.elasticsearch.common.util.set.Sets;
+
+import java.util.Collections;
+
+import static org.elasticsearch.cluster.routing.ShardRoutingState.INITIALIZING;
+import static org.elasticsearch.cluster.routing.ShardRoutingState.STARTED;
+
+/**
+ * The test case checks for the scenario when there is no data node in the cluster and only
+ * master is active. At this moment when index creation is tried, the cluster health status should
+ * change to RED
+ */
+public class NoDataNodesHealthTests extends ESAllocationTestCase {
+    /**
+     * This method specifically creates a cluster with no data nodes
+     * and a single master node
+     */
+    private ClusterState setUpClusterWithNoDataNodes() {
+
+        DiscoveryNodes node = DiscoveryNodes.builder().add(newNode("node_m", Collections.singleton(DiscoveryNode.Role.MASTER))).build();
+        MetaData metaData = MetaData.builder()
+                .put(IndexMetaData.builder("TestIndex")
+                        .settings(settings(Version.CURRENT))
+                        .numberOfShards(randomIntBetween(1, 3))
+                        .numberOfReplicas(randomIntBetween(0, 2)))
+                .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("TestIndex")).build();
+        ClusterState state = ClusterState.builder(new ClusterName("test_cluster"))
+                .nodes(node)
+                .metaData(metaData)
+                .routingTable(routingTable)
+                .build();
+        MockAllocationService service = createAllocationService();
+        state = service.reroute(state, "reroute");
+        return state;
+    }
+
+    public void testClusterHealthWithNoDataNodes() {
+        ClusterState state = setUpClusterWithNoDataNodes();
+        int dataNodes = state.nodes().getDataNodes().size();
+        int masterNodes = state.nodes().getMasterNodes().size();
+        assertTrue(dataNodes == 0);
+        assertTrue(masterNodes > 0);
+        ClusterHealthStatus clusterHealthStatus = new ClusterStateHealth(state).getStatus();
+        assertEquals(ClusterHealthStatus.RED, clusterHealthStatus);
+    }
+
+    /**
+     * The method test for scenario where we have a cluster with indices and data nodes
+     * and then all data nodes gets terminated now new index is created then
+     * all indices and cluster health should be red, but last allocation attempts of new index shards v/s
+     * old index shards should be different.
+     */
+    public void testAllocationStatusForTerminatedNodes() {
+        //creates one master and two data nodes
+        DiscoveryNodes.Builder nodeBuilder = DiscoveryNodes.builder().add(newNode("node_m", Collections.singleton(DiscoveryNode.Role.MASTER)))
```

Comment:
The master node isn't really necessary here but written like this it looks like it's important to this test. If you wanted, you could reasonably add a random number (possibly zero) non-data nodes (master nodes, coordinating-only nodes, etc) to document that the non-data nodes aren't important.